ENH: add to_offset method to Timedelta #9064

tvyomkesh opened the pull request:
Closes #9064

tvyomkesh committed:
ENH: add to_offset method to Timedelta #9064

to_offset now accepts Timedelta as input

jorisvandenbossche commented:
Another option is to ensure it is first a `pd.Timedelta` and then use `total_seconds` (this always has that, also on python 2.6).

Some other things:
- can you update the docstring of `to_offset`
- needs some more tests I think. Certainly test for both `datetime.timedelta/pandas.Timedelta` input
- I was wondering if it would be possible that eg `to_offset(pd.Timedelta('1 day'))` would return a 1 Day offset instead of 86400 Seconds offset.

cc @bjonen @cancan101

jreback commented:
your method will not handle < 1 sec if its not divisible. better to start go thru the time hierarchy and add offsets. e.g. ``pd.Timedelta('1 day 1 sec') == Day(1) + Second(1)``, you can use the ``.components`` property to get these.

tvyomkesh commented:
@jorisvandenbossche working on these
* add handling of pd.Timedelta objects. 
* update the docstring of to_offset. 
* Looking into keeping the resolution same e.g., Day(1) to Day(1) instead of Second(86400). I guess this would be taken care of if I could traverse the heirarchy and add each offset as it is (suggestion by @jreback)

@jreback my initial understanding was that converting everything into total_seconds would serve the purpose. However, going through the heirarchy and adding the offsets certainly sounds like a better way. Looking into this.

Thanks,

tvyomkesh commented:
@jorisvandenbossche I will add more tests.

tvyomkesh commented:
A question. @jreback when you say .components, are you referring to something similar to what I found in timedeltas.rst? -> "You can use the .components property to access a reduced form of the timedelta. This returns a DataFrame indexed similarly to the Series" or did you mean directly use the attributes days,hours,minutes,seconds,milliseconds,microseconds,nanoseconds? Thanks.

jreback commented:
.components of a TimeDelta is a named tuple with the data u need

jorisvandenbossche commented:
you can better use the `.components`, as the others we are probably going to change.

tvyomkesh commented:
Ok thanks. I am seeing an inconsistency though. pd.Timedelta doesn't allow nanoseconds in constructor but its components list till nanoseconds. Here is what I can reproduce.
<pre>
Python 2.7.9 (v2.7.9:648dcafa7e5f, Dec 10 2014, 10:10:46)
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pandas.tslib import Timedelta
>>> td = Timedelta(nanoseconds=1)
Traceback (most recent call last):
File "<stdin>", line 1, in <module>
File "pandas/tslib.pyx", line 1723, in pandas.tslib.Timedelta.__new__ (pandas/tslib.c:29743)
raise ValueError("cannot construct a TimeDelta from the passed arguments, allowed keywords 
are " [days, seconds, microseconds, milliseconds, minutes, hours, weeks]
>>> td=Timedelta(seconds=1)
>>> td.components._fields
('days', 'hours', 'minutes', 'seconds', 'milliseconds', 'microseconds', 'nanoseconds')
</pre>

jreback commented:
hmm
nanoseconds is not allowed as an argument to datetime.timedelta but I t should be allowed here

u can add that if u want (and a separate test)

tvyomkesh commented:
Also, if I try to add Nano() DateOffset to any other pd.DateOffset on my machine, I get this error: TypeError: ufunc add cannot use operands with types dtype('O') and dtype('<m8[ns]'). Something I need to fix in my build etc.?

Example:
<pre>
>>> from pandas import offsets
>>> d1=offsets.Second(1)
>>> d2=offsets.Milli(1)
>>> d3=offsets.Micro(1)
>>> d4=offsets.Nano(1)
>>> 5*d1+4*d2+3*d3+2*d4
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/tseries/offsets.py", line 2027, in __add__
    return _delta_to_tick(self.delta + other.delta)
TypeError: ufunc add cannot use operands with types dtype('O') and dtype('&lt;m8[ns]')
>>> 5*d1+4*d2+3*d3
&lt;5004003 * Micros&gt;
>>> 5*d1+4*d2+0*d3
&lt;5004 * Millis&gt;
>>> 5*d1+4*d2+1*d3
&lt;5004001 * Micros&gt;
</pre>

Thanks,

tvyomkesh commented:
Thanks @jreback I will club adding nanoseconds in constructor to this change.

jreback commented:
hmm adding a Nano should work
maybe not converted properly (and clearly not tested)

tvyomkesh commented:
Ok thanks, I will need to look into these more and see if I spot a fix. I guess these need to be fixed first to properly traverse hierarchy and add offsets otherwise things seem to be working fine till microseconds.

tvyomkesh commented:
@jorisvandenbossche, @jreback I have updated the PR with the following changes:
* to_offset can take either datetime.timedelta or pd.Timedelta as input
* changed doc string of the function accordingly
* fixed output time resolution to be the same as input
* added more tests

PLMK if there are better ways to get DateOffset objects from td.components._fields. I have created a local name to offset object map in the method. Will attend to review comments as they come by. In the meantime, I will look into: a) why adding Nano to DateOffset is failing and b) adding support of Nano in pd.Timedelta constructor.

jorisvandenbossche commented on pandas/tseries/frequencies.py:
```diff
@@ -278,7 +278,8 @@ def get_period_alias(offset_str):
     """
-    Return DateOffset object from string representation
+    Return DateOffset object from string representation or
+    datetime.timedelta or pandas.tslib.Timedelta objects
```
pandas.tslib.Timedelta -> pandas.Timedelta (that is how it is exposed to the users)

tvyomkesh replied:
done

jorisvandenbossche commented on pandas/tseries/frequencies.py:
```diff
@@ -298,6 +301,50 @@ def to_offset(freqstr):
+        except Exception, ex:
+            raise ValueError("Error evaluating %s. Reason: %s" % (freqstr, ex))
+
+    elif isinstance(freqstr, timedelta):
```
I think you could also just convert the `datetime.timedelta` to a `pd.Timedelta`, and then treat them the same. Maybe that is not the most efficient for `datetime.timedelta`, but then you don't need the two different codepaths

tvyomkesh replied:
done. agreed and this way to_offset will be more resilient to datetime.timedelta evolution.

jorisvandenbossche commented on pandas/tseries/tests/test_frequencies.py:
```diff
@@ -81,6 +83,96 @@ def test_to_offset_leading_zero():
+    except ValueError:
+        pass
+    else:
+        assert(False)
```
You can use `tm.assertRaises` for this, for usage see eg https://github.com/tvyomkesh/pandas/blob/vyom/firstpr/pandas/tseries/tests/test_frequencies.py#L193

tvyomkesh replied:
done

tvyomkesh commented:
From what I see, support for nano is broken in Timedelta and I am able to reproduce that in various ways. However, it seems like this ENH should should not need any further changes related to the nano fixes. Would it be a good idea to treat these two separately? I can go ahead and create a separate issue for the nano fix and continue the fix related discussions there, while this ENH can be merged independently.

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -298,6 +304,28 @@ def to_offset(freqstr):
         delta = get_offset(name) * stride
+
+    elif isinstance(freqstr, tslib.Timedelta):
+        name_to_offset_map = {'days': Day(1),
```
put this as a module level variable

tvyomkesh replied:
sounds good. done.

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -291,6 +292,11 @@ def to_offset(freqstr):
     if isinstance(freqstr, DateOffset):
         return freqstr
 
+    delta = None
```
kill this

tvyomkesh replied:
yup

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -298,6 +304,28 @@ def to_offset(freqstr):
+                              'microseconds': Micro(1),
+                              'nanoseconds': Nano(1)}
+        try:
+            for name in freqstr.components._fields:
```
add the ``delta =None here``

tvyomkesh replied:
done

jreback commented on pandas/tseries/tests/test_frequencies.py:
```diff
@@ -81,6 +83,86 @@ def test_to_offset_leading_zero():
 
 
+def test_to_offset_datetime_timedelta():
+    td = timedelta(days=1, seconds=1)
```
add the issue number as a comment here

tvyomkesh replied:
done

jreback commented on pandas/tseries/tests/test_frequencies.py:
```diff
@@ -81,6 +83,86 @@ def test_to_offset_leading_zero():
+
+
+def test_to_offset_pd_timedelta():
+    td = tslib.Timedelta(days=1, seconds=1)
```
these are virtually the same routine, so kill the first. You dont need to test both, just the ``Timedelta`` ones (its a sub-class and has identical behavior).

jorisvandenbossche replied:
sorry @tvyomkesh, as it was me who suggested to test both ..

tvyomkesh replied:
done

jreback commented:
couple of minor changes, pls rebase and squash.

jorisvandenbossche commented on pandas/tseries/frequencies.py:
```diff
@@ -291,6 +292,11 @@ def to_offset(freqstr):
+    delta = None
+
+    if isinstance(freqstr, timedelta):
+        freqstr = tslib.Timedelta(freqstr)
```
this test is not needed, you can just do `isinstance(freqstr, timedelta)` below, and then first convert to `Timedelta` (pandas.Timedelta will also be an instance of timedelta)

tvyomkesh replied:
makes sense. done

jorisvandenbossche commented on pandas/tseries/frequencies.py:
```diff
@@ -298,6 +304,28 @@ def to_offset(freqstr):
+            for name in freqstr.components._fields:
+                offset = name_to_offset_map[name]
+                stride = getattr(freqstr.components, name)
+                if stride is not None and int(stride) != 0:
```
I think stride is never None? (it will always be 0 if there are no values in that component)

jorisvandenbossche replied:
And I also think that `stride` is guaranteed to be an int, so the `int( )` is superfluous (@jreback?)

jreback replied:
yes this cannot be ``None``. Its always an int (but of course you only want to consider non-zero ones as you are doing).

tvyomkesh replied:
removed None check and int()

tvyomkesh commented:
Incorporated review comments and squashed. Created a new issue to track nano fix here. https://github.com/pydata/pandas/issues/9273. Thanks.

jreback commented:
ok, looks good. Pls add a release note referencing the original issue. You can put it in the enhancements section.

tvyomkesh commented:
@jreback done adding note

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -99,6 +99,7 @@ Enhancements
 - Added time interval selection in get_data_yahoo (:issue:`9071`)
 - Added ``Series.str.slice_replace()``, which previously raised NotImplementedError (:issue:`8888`)
 - Added ``Timestamp.to_datetime64()`` to complement ``Timedelta.to_timedelta64()`` (:issue:`9255`)
+- ``pandas.tseries.frequencies.to_offset()`` now accepts ``pandas.Timdelta`` or ``datetime.timedelta`` as input (:issue:`9064`)
```
you can just say ``Timedelta`` (spelling too)

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -298,6 +307,23 @@ def to_offset(freqstr):
+
+    elif isinstance(freqstr, timedelta):
+        delta = None
+        freqstr = tslib.Timedelta(freqstr)
```
just add the ``Timedelta`` import ``from pandas`` at the top, rather than using ``tslib`` (its the same but prefer this way of importing)

jreback commented on pandas/tseries/tests/test_frequencies.py:
```diff
@@ -17,6 +17,8 @@
 
 import pandas.util.testing as tm
+from datetime import datetime, timedelta
+import pandas.tslib as tslib
```
``from pandas import Timedelta``

jreback commented:
minor import correction, otherwise looks good. ping after you push and its green.

tvyomkesh commented:
Ok, changed imports, docstring and release note. Thanks.

jreback commented:
@tvyomkesh thanks!
